data-prepper-plugins/csv-processor/src/main/java/com/amazon/dataprepper/plugins/processor/csv/CSVProcessor.java

```diff
@@ -45,4 +83,65 @@ public boolean isReadyForShutdown() {
     public void shutdown() {
 
     }
+
+    private CsvMapper createCsvMapper() {
+        final CsvMapper mapper = new CsvMapper();
+        mapper.enable(CsvParser.Feature.WRAP_AS_ARRAY); // allows mapper to read with empty schema
+        return mapper;
+    }
+
+    private CsvSchema createCsvSchema() {
+        final char delimiterAsChar = config.getDelimiter().charAt(0); // safe due to config input validations
+        final char quoteCharAsChar = config.getQuoteCharacter().charAt(0); // safe due to config input validations
+        final CsvSchema schema = CsvSchema.emptySchema().withColumnSeparator(delimiterAsChar).withQuoteChar(quoteCharAsChar);
+        return schema;
+    }
+
+    private List<String> parseHeader(Event event, final boolean thisEventHasHeaderSource, final CsvMapper mapper,
+                                     final CsvSchema schema) {
+        if (thisEventHasHeaderSource) {
+            try {
+                final String headerUnprocessed = event.get(config.getColumnNamesSourceKey(), String.class);
+                final MappingIterator<List<String>> headerIterator = mapper.readerFor(List.class).with(schema)
+                        .readValues(headerUnprocessed);
+                // if header is empty, behaves correctly since columns are autogenerated.
+                final List<String> headerFromEventSource = headerIterator.nextValue();
+                return headerFromEventSource;
+            }
+            catch (IOException e) {
+                final List<String> emptyHeader = new ArrayList<String>();
+                return emptyHeader;
+            }
+        }
+        else if (config.getColumnNames() != null) {
+            return config.getColumnNames();
+        }
+        else {
+            final List<String> emptyHeader = new ArrayList<String>();
+            return emptyHeader;
+        }
+    }
+
+    private void putDataInEvent(Event event, final List<String> header, final List<String> data) {
+        final int dataSize = data.size();
+        final int headerSize = header.size();
+        if (dataSize <= headerSize) {
+            for (int i = 0; i < dataSize; i++) {
+                event.put(header.get(i),data.get(i));
+            }
+        } else {
+            int i = 0;
+            for (; i < headerSize; i++) {
+                event.put(header.get(i),data.get(i));
+            }
+            for (int j = i; j < dataSize; j++) {
+                event.put(generateColumn(j),data.get(j));
+            }
```

Comment:
I think this could work for `dataSize <= headerSize` too right? If `headerSize >= dataSize` then the second loop just wouldn't be executed. Cutting out the `if` would help simplify this function. 

You may also want to consider creating more informative variable names than `i` and `j`. It took me a few reads to understand what they were being used for. Maybe rewriting this function to something like this would make sense:

```
int providedHeadersColIdx = 0;
for (providedHeadersColIdx; providedHeadersColIdx < headers.size(); providedHeadersColIdx++) {
    event.put(headers.get(providedHeadersColIdx), data.get(providedHeadersColIdx));
}
for (int remainingColsIdx = providedHeadersColIdx; remainingColsIdx < data.size(); remainingColsIdx++) {
    event.put(generateColumn(remainingColsIdx),data.get(remainingColsIdx));
}
```
There might be some clever naming that you could do here. The names I have here are pretty verbose

Reply:
Great suggestion! We can do something like: 
```
int providedHeaderColIdx = 0;
for (; providedHeaderColIdx < header.size() && providedHeaderColIdx < data.size(); providedHeaderColIdx++) {
    event.put(header.get(providedHeaderColIdx), data.get(providedHeaderColIdx));
}
```
To prevent the index being out of bounds if headerSize > dataSize. Then we can cut the `if` and keep it to two `for` loops.